File: csharp/DCbor/DCbor/TagsStore.cs
namespace BlockchainCommons.DCbor;

/// <summary>
/// A delegate for summarizing tagged CBOR values as human-readable strings.
/// </summary>
/// <param name="untaggedCbor">The CBOR content without the tag.</param>
/// <param name="flat">Whether flat formatting is requested.</param>
/// <returns>A human-readable summary string.</returns>
public delegate string CborSummarizer(Cbor untaggedCbor, bool flat);

/// <summary>
/// Discriminated union for specifying which tag store to use for formatting.
/// </summary>
public abstract class TagsStoreOption
{
    private TagsStoreOption() { }

    public sealed class None : TagsStoreOption { internal static readonly None Instance = new(); }
    public sealed class Global : TagsStoreOption { internal static readonly Global Instance = new(); }
    public sealed class Custom : TagsStoreOption
    {
        public TagsStore Store { get; }
        public Custom(TagsStore store) { Store = store; }
    }

    public static TagsStoreOption NoTags => None.Instance;
    public static implicit operator TagsStoreOption(TagsStore store) => new Custom(store);

    // Default is Global
    internal static TagsStoreOption DefaultGlobal => Global.Instance;
}

/// <summary>
/// Registry that maps between CBOR tag values and human-readable names,
/// with optional summarizer delegates.
/// </summary>
public class TagsStore
{
    private readonly Dictionary<ulong, Tag> _tagsByValue = new();
    private readonly Dictionary<string, Tag> _tagsByName = new();
    private readonly Dictionary<ulong, CborSummarizer> _summarizers = new();

    public TagsStore() { }

    public TagsStore(IEnumerable<Tag> tags)
    {
        foreach (var tag in tags)
            Insert(tag);
    }

    public void Insert(Tag tag)
    {
        string name = tag.Name ?? throw new ArgumentException("Tag must have a name for registration");
        if (string.IsNullOrEmpty(name))
            throw new ArgumentException("Tag name must not be empty");

        if (_tagsByValue.TryGetValue(tag.Value, out var existing))
        {
            if (existing.Name != name)
                throw new InvalidOperationException(
                    $"Attempt to register tag: {tag.Value} '{existing.Name}' with different name: '{name}'");
        }

        _tagsByValue[tag.Value] = tag;
        _tagsByName[name] = tag;
    }

    public void InsertAll(IEnumerable<Tag> tags)
    {
        foreach (var tag in tags)
            Insert(tag);
    }

    public void SetSummarizer(ulong tagValue, CborSummarizer summarizer)
    {
        _summarizers[tagValue] = summarizer;
    }

    // --- Lookup ---

    public string? AssignedNameForTag(Tag tag)
    {
        return _tagsByValue.TryGetValue(tag.Value, out var t) ? t.Name : null;
    }

    public string NameForTag(Tag tag)
    {
        return AssignedNameForTag(tag) ?? tag.Value.ToString();
    }

    public Tag? TagForValue(ulong value)
    {
        return _tagsByValue.TryGetValue(value, out var tag) ? tag : null;
    }

    public Tag? TagForName(string name)
    {
        return _tagsByName.TryGetValue(name, out var tag) ? tag : null;
    }

    public string NameForValue(ulong value)
    {
        var tag = TagForValue(value);
        return tag?.Name ?? value.ToString();
    }

    public CborSummarizer? GetSummarizer(ulong tagValue)
    {
        return _summarizers.TryGetValue(tagValue, out var s) ? s : null;
    }

}

File: python/dcbor/src/dcbor/tags_store.py
from __future__ import annotations

import threading
from collections.abc import Callable
from typing import TYPE_CHECKING, TypeVar

from .tag import Tag, TagValue

if TYPE_CHECKING:
    from .cbor import CBOR

T = TypeVar("T")


CBORSummarizer = Callable[["CBOR", bool], str]


class TagsStore:
    """Registry mapping CBOR tag numbers to named Tag objects and optional diagnostic summarizers."""

    __slots__ = ("_tags_by_value", "_tags_by_name", "_summarizers")

    def __init__(self, tags: list[Tag] | None = None) -> None:
        self._tags_by_value: dict[TagValue, Tag] = {}
        self._tags_by_name: dict[str, Tag] = {}
        self._summarizers: dict[TagValue, CBORSummarizer] = {}
        if tags:
            for tag in tags:
                self._do_insert(tag)

    def insert(self, tag: Tag) -> None:
        self._do_insert(tag)

    def insert_all(self, tags: list[Tag]) -> None:
        for tag in tags:
            self._do_insert(tag)

    def set_summarizer(self, tag_value: TagValue, summarizer: CBORSummarizer) -> None:
        self._summarizers[tag_value] = summarizer

    def _do_insert(self, tag: Tag) -> None:
        name = tag.name
        if name is None or name == "":
            raise ValueError("Tag must have a non-empty name")
        old = self._tags_by_value.get(tag.value)
        if old is not None:
            old_name = old.name
            if old_name != name:
                raise ValueError(
                    f"Attempt to register tag: {tag.value} '{old_name}' "
                    f"with different name: '{name}'"
                )
        self._tags_by_value[tag.value] = tag
        self._tags_by_name[name] = tag

    def assigned_name_for_tag(self, tag: Tag) -> str | None:
        found = self._tags_by_value.get(tag.value)
        if found is not None:
            return found.name
        return None

    def name_for_tag(self, tag: Tag) -> str:
        name = self.assigned_name_for_tag(tag)
        if name is not None:
            return name
        return str(tag.value)

    def tag_for_value(self, value: TagValue) -> Tag | None:
        return self._tags_by_value.get(value)

    def tag_for_name(self, name: str) -> Tag | None:
        return self._tags_by_name.get(name)

    def name_for_value(self, value: TagValue) -> str:
        tag = self.tag_for_value(value)
        if tag is not None and tag.name is not None:
            return tag.name
        return str(value)

    def summarizer(self, tag_value: TagValue) -> CBORSummarizer | None:
        return self._summarizers.get(tag_value)


_global_lock = threading.Lock()
_global_tags: TagsStore | None = None


def _get_global_tags() -> TagsStore:
    global _global_tags
    if _global_tags is None:
        with _global_lock:
            if _global_tags is None:
                _global_tags = TagsStore()
    return _global_tags


def with_tags(action: Callable[[TagsStore], T]) -> T:
    return action(_get_global_tags())


def with_tags_mut(action: Callable[[TagsStore], T]) -> T:
    with _global_lock:
        return action(_get_global_tags())


TAG_DATE: TagValue = 1
_TAG_NAME_DATE = "date"


def tags_for_values(values: list[TagValue]) -> list[Tag]:
    store = _get_global_tags()
    result: list[Tag] = []
    for value in values:
        tag = store.tag_for_value(value)
        if tag is not None:
            result.append(tag)
        else:
            result.append(Tag.with_value(value))
    return result


def register_tags_in(store: TagsStore) -> None:
    from .date import Date

    store.insert(Tag(TAG_DATE, _TAG_NAME_DATE))

    def date_summarizer(cbor: CBOR, flat: bool) -> str:
        return str(Date.from_untagged_cbor(cbor))

    store.set_summarizer(TAG_DATE, date_summarizer)


def register_tags() -> None:
    """Register the standard dcbor tags (date, etc.) in the global TagsStore."""
    with_tags_mut(lambda store: register_tags_in(store))
